Mathlib/RingTheory/PowerSeries/Inverse.lean

```diff
@@ -208,16 +216,79 @@ set_option linter.uppercaseLean3 false in
 #align power_series.C_inv PowerSeries.C_inv
 
 @[simp]
-theorem X_inv : (X : PowerSeries k)⁻¹ = 0 :=
+theorem X_inv : (X : k⟦X⟧)⁻¹ = 0 :=
   MvPowerSeries.X_inv _
 set_option linter.uppercaseLean3 false in
 #align power_series.X_inv PowerSeries.X_inv
 
 @[simp]
-theorem smul_inv (r : k) (φ : PowerSeries k) : (r • φ)⁻¹ = r⁻¹ • φ⁻¹ :=
+theorem smul_inv (r : k) (φ : k⟦X⟧) : (r • φ)⁻¹ = r⁻¹ • φ⁻¹ :=
   MvPowerSeries.smul_inv _ _
 #align power_series.smul_inv PowerSeries.smul_inv
 
+/-- `firstUnitCoeff` is the non-zero coefficient whose index is `f.order`, seen as a unit of the
+  field. It is obtained using `divided_by_X_pow_order`, defined in `PowerSeries.Order`-/
+def firstUnitCoeff {f : k⟦X⟧} (hf : f ≠ 0) : kˣ := by
+  let d := f.order.get (order_finite_iff_ne_zero.mpr hf)
+  have f_const : coeff k d f ≠ 0 := by apply coeff_order
+  have : Invertible (constantCoeff k (divided_by_X_pow_order hf)) := by
+    apply invertibleOfNonzero
+    convert f_const using 1
+    rw [← coeff_zero_eq_constantCoeff, ← zero_add d]
+    convert (coeff_X_pow_mul (exists_eq_mul_right_of_dvd (X_pow_order_dvd
+      (order_finite_iff_ne_zero.mpr hf))).choose d 0).symm
+    exact (self_eq_X_pow_order_mul_divided_by_X_pow_order hf).symm
+  exact unitOfInvertible (constantCoeff k (divided_by_X_pow_order hf))
+
+/-- `Inv_divided_by_X_pow_order` is the inverse of the element obtained by diving a non-zero power
+series by the largest power of `X` dividing it. Useful to create a term of type `Units`, done in
+`Unit_divided_by_X_pow_order` -/
+def Inv_divided_by_X_pow_order {f : k⟦X⟧} (hf : f ≠ 0) : k⟦X⟧ :=
+  invOfUnit (divided_by_X_pow_order hf) (firstUnitCoeff hf)
+
+@[simp]
+theorem Inv_divided_by_X_pow_order_rightInv {f : k⟦X⟧} (hf : f ≠ 0) :
+    divided_by_X_pow_order hf * Inv_divided_by_X_pow_order hf = 1 :=
+  mul_invOfUnit (divided_by_X_pow_order hf) (firstUnitCoeff hf) rfl
+
+@[simp]
+theorem Inv_divided_by_X_pow_order_leftInv {f : k⟦X⟧} (hf : f ≠ 0) :
+    (Inv_divided_by_X_pow_order hf) * (divided_by_X_pow_order hf) = 1 := by
+  rw [mul_comm]
+  exact mul_invOfUnit (divided_by_X_pow_order hf) (firstUnitCoeff hf) rfl
+
+open scoped Classical
```

Comment:
```suggestion
```
Instead of this you can use `[Decidable (f = 0)]` in the definitions, and start the proofs by `Classical`.

Reply:
Done